src/feed.sol

```diff
@@ -0,0 +1,148 @@
+    function file(bytes32 what, bytes32 nftID_, uint maturityDate_) public {
+        if (what == "maturityDate") {
+            maturityDate[nftID_] = uniqueDayTimestamp(maturityDate_);
+        } else { revert("unknown config parameter");}
+    }
+
+    /// Ceiling Implementation
+    function borrow(uint loan, uint amount) external auth {
+        uint normalizedDay = uniqueDayTimestamp(now);
+
+        // ceiling check uses existing loan debt
+        require(ceiling(loan) >= safeAdd(pile.debt(loan), amount), "borrow-amount-too-high");
+
+        bytes32 nftID_ = nftID(loan);
+        // calculate future cash flow
+        uint maturityDate_ = maturityDate[nftID_];
+
+
+        if (dateBucket[maturityDate_] == 0) {
+            addToLinkedList(maturityDate_);
+        }
+
+        // calculate future value of the loan and add it to the bucket
+        dateBucket[maturityDate_] = safeAdd(dateBucket[maturityDate_],
+            rmul(rpow(pile.loanRates(loan),  safeSub(maturityDate_, normalizedDay), ONE), amount));
+
+    }
+
+    function addToLinkedList(uint maturityDate_) internal {
+        if (firstBucket == 0) {
+            firstBucket = maturityDate_;
+            nextBucket[maturityDate_] = NullDate;
+            return;
+        }
+
+        // new bucket before first one
+        if (maturityDate_ < firstBucket) {
+            nextBucket[maturityDate_] = firstBucket;
+            firstBucket = maturityDate_;
+            return;
+        }
+
+        // find previous bucket
+        uint prev = maturityDate_;
+        while(nextBucket[prev] == 0) {prev = prev - 1 days;}
+
+        nextBucket[maturityDate_] = nextBucket[prev];
+        nextBucket[prev] = maturityDate_;
+    }
+
+    function repay(uint loan, uint amount) external auth {
+        // todo
+        // remove from FV
+        // remove from linked list if bucket is zero
+
+    }
+
+    /// deprecated
+    /// returns the NAV (net asset value) of the pool
+    /// only for performance comparing
+    /// todo old implementation will be removed
+    function navOld() public view returns(uint) {
+        uint normalizedDay = uniqueDayTimestamp(now);
+        uint sum = 0;
+
+        // current implementation ignores overdue nfts
+        for (uint i = 0;i <= maxDays; i=i + 1 days) {
+            if(dateBucket[normalizedDay + i] != 0) {
+                sum += rdiv(dateBucket[normalizedDay + i], rpow(discountRate,  i, ONE));
+            }
+        }
+        return sum;
+    }
+
+    /// returns the NAV (net asset value) of the pool
+    function nav() public view returns(uint) {
+        uint normalizedDay = uniqueDayTimestamp(now);
+        uint sum = 0;
+
+        uint currDate = normalizedDay;
+
+        // we could remember the next bucket and only perform the search
+        // if now > remembered bucket
+        // con: we would modify the state and its not view anymore
+        while(nextBucket[currDate] == 0) { currDate = currDate + 1 days; }
```

Comment:
the optimization might look as follow:
it would avoid the `while` loop iteration every time
```javascript
      if(now > nextFutureBucket || nextFutureBucket == 0) {
            while(nextBucket[currDate] == 0) { currDate = currDate + 1 days; }
            nextFutureBucket = currDate;
        } else {
            currDate = nextFutureBucket;
        }
```